add "Scaling-Out Development Environment" by Tanjim Hossain (#64)
metalbear-co/metalbear.co: content/en/blog/scaling-out-development-environment/index.md

---

> title: "Scaling-Out Development Environment"

Reviewer: ```suggestion
title: "Scaling Out Development Environments"
```

description: "How mirrord helps to alleviate the pain of dev-prod parity"

> lead: "mirrord to close the gap between Dev & Prod"

Reviewer: ```suggestion
lead: "Using mirrord to close the gap between Dev & Prod"
```

tags:
  - metalbear
  - mirrord
  - kubernetes
  - devops
  - devenv
  - devx
  - devcontainers
  - vscode
  - cloudflared
  - tailscale
  - k3s
  - minikube

> date: 2023-12-21T0:00:00+00:00

Reviewer: ```suggestion
date: 2024-01-02T0:00:00+00:00
```

lastmod: 2023-12-21T0:00:00+00:00

> draft: false

Reviewer: ```suggestion
draft: false
weight: 50
```

Reviewer: (This is needed for the post to show up in the correct place)

images: ["cover.webp"]
contributors: ["Tanjim Hossain"]
---

> Cross-posted from [https://audacioustux.com/Scaling-Out-Development-Environment-ceb823260a1f4a10b717a87e998480de](https://audacioustux.com/Scaling-Out-Development-Environment-ceb823260a1f4a10b717a87e998480de)

Once upon a time, there was LAMP Stack… **L**inux, **A**pache, **M**ysql, **P**hp/**P**ython/**P**erl - That’s what I’ve used to start my Software Engineering journey. Things were simple, in a sense that there were less “options” to choose from. But today, I have to deploy multiple types of databases, event store + processing pipeline, and other behemoths to work on my application locally.

Well, maybe I can use docker-compose to keep things clean and tidy, re-producible for the other team members, but what about the compute resources usage? I don’t know about the M2/M3 users out there, but my Linux laptop either halts or burns my lap.

> Soo, how ‘bout I offload some of the compute intensive stuffs to remote machine, or maybe cloud? Although, that introduces some other factors like cloud cost, internet connectivity…, what if you have an old machine laying around in your house? That’s exactly what I did at some point.

Reviewer: ```suggestion
So, how ‘bout I offload some of the compute intensive stuff to a remote machine, or maybe the cloud? Although, that introduces some other factors like cloud cost, internet connectivity… what if you have an old machine laying around in your house? That’s exactly what I did at some point.
```

But there’re still questions and issues…

> 1. **Re-producibility:** So, you have all the containers with their configurations specified in a docker-compose file, great! All the platform needs, including the other services you have as dependency (if you’re into ~~micro~~-services) are now accessible to your application, and re-producible by simple `docker compose up` Boom, done! But what about the CLI tools, desktop apps, IDE extensions, and so-on? Maybe you or your teammates each have multiple machines for their development work. It’s not only the platform, or the application you’re working on should be containerized, but also the whole **Dev Environment!**

Reviewer: ```suggestion
1. **Reproducibility:** So, you have all the containers with their configurations specified in a docker-compose file, great! All the platform needs, including the other services you have as dependencies (if you’re into ~~micro~~-services) are now accessible to your application, and reproducible with a simple `docker compose up`. Boom, done! But what about the CLI tools, desktop apps, IDE extensions, and so on? Maybe you or your teammates each have multiple machines for their development work. It’s not only the platform, or the application you’re working on that should be containerized, but also the whole **Dev Environment!**
```

> 2. **Dev-Prod Parity:** Are you willing to sacrifice the “Twelve-Factor App Methodology” - [Twelve-Factor App methodology](https://en.wikipedia.org/wiki/Twelve-Factor_App_methodology)

Reviewer: ```suggestion
2. **Dev-Prod Parity:** Are you willing to sacrifice the [“Twelve-Factor App Methodology”] (https://en.wikipedia.org/wiki/Twelve-Factor_App_methodology)
```

Reviewer: and delete the newline before "because" please

>     because of the constraints you have in your local dev setup? Not so ideal, right? In this day & age, typically your app should be ready for multi-instance deployment, or even multi-region for HA. It’s hard to anticipate what would happen in that kind-of production dynamic, if you only ever deployed a single instance of the app in docker compose manifest. Also, if your target deployment platform is Kubernetes (EKS, K3s, AKE, GKE, OKE, CIVO etc.), why docker compose!? why docker!? why container!? ahem… ok, I’ll explain the last one, I promise 🤞

Reviewer: ```suggestion
    because of the constraints you have in your local dev setup? Not so ideal, right? In this day & age, typically your app should be ready for multi-instance deployment, or even multi-region for HA. It’s hard to anticipate what would happen in that kind of production dynamic, if you only ever deployed a single instance of the app in a docker compose manifest. Also, if your target deployment platform is Kubernetes (EKS, K3s, AKE, GKE, OKE, CIVO etc.), why docker compose!? why docker!? why containers!? ahem… ok, I’ll explain the last one, I promise 🤞
```

> 3. **CNCF Landscape & Kubernetes:** Have you seen the CNCF Landscape? [Cloud Native Landscape](https://landscape.cncf.io/)

Reviewer: ```suggestion
3. **CNCF Landscape & Kubernetes:** Have you seen the [CNCF Landscape](https://landscape.cncf.io/)?
```

>     Yah, it’s Massive! But my point is, many of those tools are not very easy to configure in simple compose files. They may not have any detail of the required setup, other than just a helm or Kubernetes manifests. That means, you’ll face some issues if your application architecture is not too simple, or the tools you use are from recent generation.

Reviewer: ```suggestion
    Yah, it’s massive! But my point is, many of those tools are not very easy to configure in simple compose files. They may not have any detail of the required setup, other than just a Helm or Kubernetes manifests. That means you’ll face some issues if your application architecture is not too simple, or the tools you use are from a recent generation.
```

> 4. **Change-Build-Test (CBT) Performance:** How do you containerize your app for dev environment? Maybe you don’t do that, instead use port bindings to expose all the dependencies you have on other services, and access them from your app, running as a local process. So, there’re 4 options here:

Reviewer: ```suggestion
4. **Change-Build-Test (CBT) Performance:** How do you containerize your app for the dev environment? Maybe you don’t do that, and instead use port bindings to expose all the dependencies you have on other services, and access them from your app, running as a local process. So, there’re 4 options here:
```

>     1. Build your app locally, copy the final artifact to container image (using COPY in Dockerfile for example) - **Cons:** Increased container build context sync time - due to the nature of build artifacts (e.g., large in size, hard to sync only the delta of the change) **Pros:** Potentially re-uses the local artifacts that your IDE may have already created, reduced redundant work!

Reviewer: ```suggestion
    1. Build your app locally, copy the final artifact to a container image (using COPY in Dockerfile for example) - **Cons:** Increased container build context sync time - due to the nature of build artifacts (e.g., large in size, hard to sync only the delta of the change) **Pros:** Potentially re-uses the local artifacts that your IDE may have already created, reducing redundant work!
```

>     1. Build your app locally, copy the final artifact to a container image (using COPY in Dockerfile for example) - **Cons:** Increased container build context sync time - due to the nature of build artifacts (e.g., large in size, hard to sync only the delta of the change) **Pros:** Potentially re-uses the local artifacts that your IDE may have already created, reducing redundant work!

Reviewer: ```suggestion
    1. Build your app locally, copy the final artifact to a container image (using COPY in Dockerfile for example).<br> **Cons:** Increased container build context sync time - due to the nature of build artifacts (e.g., large in size, hard to sync only the delta of the change) <br>**Pros:** Potentially re-uses the local artifacts that your IDE may have already created, reducing redundant work!
```

Reviewer: We have to sort out our CSS, but until then we need an explicit <br> here for a newline 🤦

>     2. Sync codebase (mount) inside container. In this case, the hot-reload kind of stuffs are intended to happen inside the container. This may not work though in windows machines due to some issues with filesystem event propagating to the container from host (not sure if the issue is now fixed or not). **Cons:** significant amount of redundant work by the IDE, and process (e.g., compiler, bundler, etc.) inside the container on each code change. **Pros:** only the particular changed file can be synced, greatly improves feedback loop in case the container runs on a remote host.

Reviewer: ```suggestion
    2. Sync codebase (mount) inside container. In this case, the hot-reload kind of stuff is intended to happen inside the container. This may not work in Windows machines though, due to some issues with filesystem event propagation to the container from the host (not sure if the issue is now fixed or not). **Cons:** significant amount of redundant work by the IDE, and process (e.g., compiler, bundler, etc.) inside the container on each code change. **Pros:** only the particular changed file can be synced, greatly improves the feedback loop in case the container runs on a remote host.
```

>     2. Sync codebase (mount) inside container. In this case, the hot-reload kind of stuff is intended to happen inside the container. This may not work in Windows machines though, due to some issues with filesystem event propagation to the container from the host (not sure if the issue is now fixed or not). **Cons:** significant amount of redundant work by the IDE, and process (e.g., compiler, bundler, etc.) inside the container on each code change. **Pros:** only the particular changed file can be synced, greatly improves the feedback loop in case the container runs on a remote host.

Reviewer: ```suggestion
    2. Sync codebase (mount) inside container. In this case, the hot-reload kind of stuff is intended to happen inside the container. This may not work in Windows machines though, due to some issues with filesystem event propagation to the container from the host (not sure if the issue is now fixed or not). <br>**Cons:** significant amount of redundant work by the IDE, and process (e.g., compiler, bundler, etc.) inside the container on each code change. <br>**Pros:** only the particular changed file can be synced, greatly improves the feedback loop in case the container runs on a remote host.
```

>     3. Using Tilt, Scaffold, etc. Tilt for example can be used to combine of both (a) and (b) [https://docs.tilt.dev/example_java.html](https://docs.tilt.dev/example_java.html) **Cons:** debugging can be hard in certain language, runtimes - due to the fact that the application process runs inside the container. **Pros:** Fast enough feedback loop / CBT cycle, whether the deployment target resides on the same machine or remote (remote k8s cluster or docker context)

Reviewer: ```suggestion
    3. Using Tilt, Skaffold, etc. Tilt for example can be used to combine of both (a) and (b) [https://docs.tilt.dev/example_java.html](https://docs.tilt.dev/example_java.html) **Cons:** debugging can be hard in certain languages and runtimes - due to the fact that the application process runs inside the container. **Pros:** Fast enough feedback loop / CBT cycle, whether the deployment target resides on the same machine or remote (remote k8s cluster or docker context)
```

>     3. Using Tilt, Skaffold, etc. Tilt for example can be used to combine of both (a) and (b) [https://docs.tilt.dev/example_java.html](https://docs.tilt.dev/example_java.html) **Cons:** debugging can be hard in certain languages and runtimes - due to the fact that the application process runs inside the container. **Pros:** Fast enough feedback loop / CBT cycle, whether the deployment target resides on the same machine or remote (remote k8s cluster or docker context)

Reviewer: ```suggestion
    3. Using Tilt, Skaffold, etc. Tilt for example can be used to combine of both (a) and (b) [https://docs.tilt.dev/example_java.html](https://docs.tilt.dev/example_java.html) <br>**Cons:** debugging can be hard in certain languages and runtimes - due to the fact that the application process runs inside the container. <br>**Pros:** Fast enough feedback loop / CBT cycle, whether the deployment target resides on the same machine or remote (remote k8s cluster or docker context)
```

>     3. Using Tilt, Skaffold, etc. Tilt for example can be used to combine of both (a) and (b) [https://docs.tilt.dev/example_java.html](https://docs.tilt.dev/example_java.html) **Cons:** debugging can be hard in certain languages and runtimes - due to the fact that the application process runs inside the container. **Pros:** Fast enough feedback loop / CBT cycle, whether the deployment target resides on the same machine or remote (remote k8s cluster or docker context)

Reviewer: ```suggestion
    3. Using Tilt, Skaffold, etc. Tilt for example [can be used to combine of both (a) and (b)](https://docs.tilt.dev/example_java.html) **Cons:** debugging can be hard in certain languages and runtimes - due to the fact that the application process runs inside the container. **Pros:** Fast enough feedback loop / CBT cycle, whether the deployment target resides on the same machine or remote (remote k8s cluster or docker context)
```

>     4. No containerization. **Cons:** hard to match production environment **Pros:** bare-minimal overhead

Reviewer: ```suggestion
    4. No containerization. <br>**Cons:** hard to match production environment <br>**Pros:** bare-minimal overhead
```

> I think we should talk / discuss more about the development environments of our software, it’s no-less important than the production environment, and should strive to improve it as much as possible. In the beginning, it may apparently feel like waste of time… but in the end, it may still look the same 😅, because the way we take things for granted, and tends to not be seeing how past actions leads to the present. As I have to work in DevOps / Platform Engineering stuffs, I had to solve some of dev environment issues not as “optional” but “mandatory” things. In the rest of this post, I’ll introduce or explain some of the tools and approaches that made my life much easier. Also, I think most of it can be applicable in a whole organization scale and save a lot of paid employee hours!

Reviewer: ```suggestion
I think we should talk / discuss more about the development environments of our software, it’s no less important than the production environment, and we should strive to improve it as much as possible. In the beginning, it may feel like waste of time… but in the end, it may still look the same 😅, because the way we take things for granted, and tends to not be seeing how past actions leads to the present. As I have to work in DevOps / Platform Engineering stuffs, I had to solve some of dev environment issues not as “optional” but “mandatory” things. In the rest of this post, I’ll introduce or explain some of the tools and approaches that made my life much easier. Also, I think most of it can be applicable in a whole organization scale and save a lot of paid employee hours!
```

> I think we should talk / discuss more about the development environments of our software, it’s no less important than the production environment, and we should strive to improve it as much as possible. In the beginning, it may feel like waste of time… but in the end, it may still look the same 😅, because the way we take things for granted, and tends to not be seeing how past actions leads to the present. As I have to work in DevOps / Platform Engineering stuffs, I had to solve some of dev environment issues not as “optional” but “mandatory” things. In the rest of this post, I’ll introduce or explain some of the tools and approaches that made my life much easier. Also, I think most of it can be applicable in a whole organization scale and save a lot of paid employee hours!

Reviewer: ```suggestion
I think we should talk / discuss more about the development environments of our software, it’s no less important than the production environment, and we should strive to improve it as much as possible. In the beginning, it may feel like waste of time… but in the end, it may still look the same 😅, because the way we take things for granted, and tend to not be seeing how past actions leads to the present. As I have to work in DevOps / Platform Engineering stuff, I had to solve some of dev environment issues not as “optional” but “mandatory” things. In the rest of this post, I’ll introduce or explain some of the tools and approaches that made my life much easier. Also, I think most of it can be applicable in a whole organization scale and save a lot of paid employee hours!
```

## Kubernetes for Dev

---

> Kubernetes, without any doubts, have a steep learning curve. Not everyone involved in a project should be well-versed in 100s of CRDs and yaml manifests!

Reviewer: ```suggestion
Kubernetes, without any doubts, has a steep learning curve. Not everyone involved in a project should be well-versed in 100s of CRDs and yaml manifests!
```

Author: Damn, that's a lot of typos, my bad, Thanks! LGTM.

> {{<figure src="kelsey-hightower-tweet.png" alt="Kubernetes is a platform for building platforms. It's a better place to start; not the endgame" height="100%" width="100%">}}

Reviewer: ```suggestion
{{< tweet user="kelseyhightower" id="935252923721793536" >}}
```

Tilt helps you to take your application closer to the platform, on the other hand, mirrord brings the platform closer to the application. As a result, it becomes much easier to debug application, integrate with IDEs, avoid redundant compute usage.

You may use the vscode extension for mirrord, or the CLI directly.

- **Mirrord CLI example**

    Here, If I run `task cargo:dev` mirrord will steal all the traffic coming to the target deployment and forward them to the application (written in rust in this case). Any outgoing request from the application will be resolved in context of the deployment. That means cluster local IP addresses and DNS resolves will work as expected, and all the environment variables passed to the deployment is available to the process running locally!

    ```yaml
    # https://taskfile.dev/usage/
    version: "3"
    
    tasks:
      # cargo check for failing fast on build time errors
      cargo:check:
        cmds:
          - touch .trigger # add .trigger in .gitignore
          - cargo watch -x check -s 'touch .trigger'
        internal: true

>       cargo:watch:

Reviewer: ```suggestion
      cargo:watch :
```

Reviewer: The emoji thing is an [issue with Hugo](https://github.com/gohugoio/hugo/issues/7332), the framework we use for our website. I think adding a space here is the cleanest solution (though still pretty ugly)

Author: good enough workaround, I think.
